File: algorithm/diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int total = 0;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return total;
    }
    
    int height(TreeNode* root) {
        if (root == nullptr)
            return 0;
        int left = height(root->left),
            right = height(root->right);
        total = max(total, left + right);
        return max(left, right) + 1;
    }
    
};

File: algorithm/diameter-of-binary-tree/diameter-of-binary-tree.go
package diameter_of_binary_tree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func diameterOfBinaryTree(root *TreeNode) int {
	maxLength := 0
	traverse(root, &maxLength)
	return maxLength
}

func traverse(root *TreeNode, maxLength *int) int {
	if root == nil {
		return 0
	}

	len1 := traverse(root.Left, maxLength)
	len2 := traverse(root.Right, maxLength)
	*maxLength = getMax(*maxLength, len1+len2)
	return getMax(len1, len2) + 1
}

func getMax(a, b int) int {
	if a > b {
		return a
	}
	return b
}
